src/platform_impl/web/web_sys/canvas/pointer_handler.rs

```diff
@@ -105,11 +115,113 @@ impl PointerHandler {
+                }
+
+                handler(
+                    event.pointer_id(),
+                    touch_physical_position(&event, &canvas),
+                    Force::Normalized(event.pressure() as f64),
+                );
+            },
+        ));
+    }
+
+    pub fn on_touch_down<F>(&mut self, canvas_common: &super::Common, mut handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        let canvas = canvas_common.raw.clone();
+        self.on_touch_down = Some(canvas_common.add_event(
+            "pointerdown",
+            move |event: PointerEvent| {
+                if event.pointer_type() != "touch" {
+                    return;
+                }
+
+                handler(
+                    event.pointer_id(),
+                    touch_physical_position(&event, &canvas),
+                    Force::Normalized(event.pressure() as f64),
+                );
+            },
+        ));
+    }
+
+    pub fn on_touch_up<F>(&mut self, canvas_common: &super::Common, mut handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        let canvas = canvas_common.raw.clone();
+        self.on_touch_up = Some(canvas_common.add_event(
+            "pointerup",
+            move |event: PointerEvent| {
+                if event.pointer_type() != "touch" {
+                    return;
+                }
+
+                handler(
+                    event.pointer_id(),
+                    touch_physical_position(&event, &canvas),
+                    Force::Normalized(event.pressure() as f64),
+                );
+            },
+        ));
+    }
+
+    pub fn on_touch_cancel<F>(&mut self, canvas_common: &super::Common, mut handler: F)
+    where
+        F: 'static + FnMut(i32, PhysicalPosition<f64>, Force),
+    {
+        let canvas = canvas_common.raw.clone();
+        self.on_touch_cancel = Some(canvas_common.add_event(
+            "pointercancel",
+            move |event: PointerEvent| {
+                if event.pointer_type() != "touch" {
+                    return;
+                }
+
+                handler(
+                    event.pointer_id(),
+                    touch_physical_position(&event, &canvas),
+                    Force::Normalized(event.pressure() as f64),
+                );
+            },
+        ));
+    }
+
     pub fn remove_listeners(&mut self) {
         self.on_cursor_leave = None;
         self.on_cursor_enter = None;
         self.on_cursor_move = None;
         self.on_pointer_press = None;
         self.on_pointer_release = None;
+        self.on_touch_move = None;
+        self.on_touch_down = None;
+        self.on_touch_up = None;
+        self.on_touch_cancel = None;
+    }
+}
+
+fn touch_physical_position(
+    event: &PointerEvent,
+    canvas: &HtmlCanvasElement,
+) -> PhysicalPosition<f64> {
+    // not window scale factor here because we can modify the scale factor.
```

Comment:
I'm not quite sure what you mean by this.

Reply:
I'd probably phrase it `// Calculate the scale factor every time, since it's possible for the canvas's width and height attributes to get out of sync with its actual HTML size.` (I think that's what you meant?)

Reply:
@Liamolucko Yes, yours is fine.